### ISSUE TITLE
Persist vectorStore to a file for use across sessions

### ISSUE THREAD

**Issue by craigbehnke** [logancyang/obsidian-copilot#46]
# Description

Right now, it appears that the vector store is generated 1) whenever the user selects the 'Use Active Note as Context' button for a note that has either not been used as context this session or has been updated, and 2) whenever the session is reset or reloaded. ([Link to code](https://github.com/logancyang/obsidian-copilot/blob/379dd90caeb6dd6825ebf72129923d4c58b9a69d/src/components/Chat.tsx#LL152C13-L152C20))

Persisting the vector store map to the vault would enable embeddings to be reused to reduce the number of hits to the embeddings API.

# Suggested Fix

Here's my suggestion on how this could be addressed; feel free to completely ignore everything after this point. (I am, after all, just an internet stranger)

Since the vector store currently is a [Map<string, VectorStore>](https://github.com/logancyang/obsidian-copilot/blob/379dd90caeb6dd6825ebf72129923d4c58b9a69d/src/chainFactory.ts#LL47C33-L47C57) object, it should be possible to save it to a JSON file from the [onunload()](https://github.com/obsidianmd/obsidian-sample-plugin/blob/2aee08d3d5e950cc6c121904f5f1a918ee6282f9/main.ts#L81) method in main.ts, and then load it in [onload()](https://github.com/logancyang/obsidian-copilot/blob/379dd90caeb6dd6825ebf72129923d4c58b9a69d/src/main.ts#L39) in the same file.

However, persisting the embeddings would also cause the map to bloat to an absurd degree, so a garbage collection function should also be added that removes vector stores that are out of date. A general outline for this function could be:
- Compile a list of hashes for all notes in the vault (List 1)
    - Since this would be on-device, there would be no info leakage
- Get a list of the hashes used as keys in the map (List 2)
- Create a third list of hashes that are in List 2 but not in List 1 (i.e. it is a hash of a note that no longer exists)
- Remove all vector stores with their keys in List 3

I'm not sure where the best place to run this would be; my gut call is that it could be triggered as part of the [Active Context button](https://github.com/logancyang/obsidian-copilot/blob/379dd90caeb6dd6825ebf72129923d4c58b9a69d/src/components/Chat.tsx#L119), but that might cause lag or latency for larger vaults.

A way to limit this latency might be to update the key format to include the note title (such as {title}-{hash}) so that List 1 could be filtered down to only the relevant notes prior to calculating out all of the hashes. And then if the title changes, or a note is deleted, that vector store can be marked for deletion right away. This would limit the scope of garbage collection to only the notes that have been used as Active Context in the past.

---

By the way, this project is absolutely awesome. Not only does it work great out of the box, but the code is also much more readable than some similar plugins. So kudos, and thanks for building this!

**Comment by logancyang** [comment]
Thanks for the detailed description and suggestions!

I'm actually looking into this exact thing. Currently blocked by the fact that `JSON.stringify` and `JSON.parse` actually lose the `asRetriever` function: https://github.com/hwchase17/langchainjs/issues/1468.

It appears that `MemoryVectorStore` is the only vector search db that works in-memory with LangchainJS. Once I find a way to do serialization/deserialization of `MemoryVectorStore` correctly, data persistence will be supported! Please feel free to share any suggestions if you know how to do this without changes from the LangchainJS side.

**Comment by logancyang** [comment]
@craigbehnke Just tried and this is not working. LangchainJS people responded to that issue just now, saying it's in their plan. But it would be great to have a temporary hack that works.

**Comment by wenlzhang** [comment]
Not sure if it helps, the following two plugins also use embedding to avoid repeatedly embed old notes:
- https://github.com/exoascension/vault-chat
- https://github.com/M7mdisk/obsidian-gpt

**Comment by logancyang** [comment]
@wenlzhang thanks for the pointers. It's not hard to implement a local vector store myself but I'm thinking of completely new setups via docker container for more powerful future features. So unless langchainJS is giving me this functionality for free, I'm putting this on hold for now and will probably shift away from langchain gradually.

**Comment by logancyang** [comment]
New release v2.4.2. Implemented the local persistence with PouchDB, added a command to clear the entire vector store, and a TTL setting.

More advanced garbage collection will be added later as needed. Closing this one for now.
